components/ws-manager/pkg/manager/monitor.go

```diff
@@ -377,42 +372,22 @@ func (m *Monitor) actOnPodEvent(ctx context.Context, status *api.WorkspaceStatus
 	}
 
 	if status.Phase == api.WorkspacePhase_STOPPING {
-		// This may be the last pod-based status we'll ever see for this workspace, so we must store it in the
-		// plis config map which in turn will trigger the status update mechanism. Because we serialize events
-		// for each workspace, the cfgmap event won't be handled before this function finishes.
-		annotations := make([]*annotation, 0)
-		if v, neverReady := pod.Annotations[workspaceNeverReadyAnnotation]; neverReady {
-			// workspace has never been ready, mark it as such.
-			// Note: this is the reason we're using a never-ready flag instead of a "positive ready" flag.
-			//       If we don't copy this flag for some reason (e.g. because we missed the stopping event),
-			//       we'll still think the workspace ready, i.e. it's ready by default.
-			annotations = append(annotations, addMark(workspaceNeverReadyAnnotation, v))
-		}
-
-		err := m.manager.patchPodLifecycleIndependentState(ctx, status.Id, func(plis *podLifecycleIndependentState) (needsUpdate bool) {
-			needsUpdate = false
-
-			if !reflect.DeepEqual(plis.LastPodStatus, status) {
-				plis.LastPodStatus = status
-				needsUpdate = true
-			}
-
-			hostIP := pod.Status.HostIP
-			if hostIP != "" && plis.HostIP != hostIP {
-				plis.HostIP = hostIP
-				needsUpdate = true
-			}
-
-			if plis.StoppingSince == nil {
-				t := time.Now().UTC()
-				plis.StoppingSince = &t
-				needsUpdate = true
+		if status.Conditions.FinalBackupComplete == api.WorkspaceConditionBool_TRUE {
+			// we've disposed already - try to remove the finalizer and call it a day
+			return m.manager.modifyFinalizer(ctx, workspaceID, gitpodFinalizerName, false)
+		}
+
+		var terminated bool
+		for _, c := range wso.Pod.Status.ContainerStatuses {
+			if c.Name == "workspace" {
+				// Note: sometimes container don't enter `terminated`, but `waiting`. The processes are stopped nonetheless,
+				//       and we should be running the backup. The only thing that keeps the pod alive, is our finalizer.
+				terminated = c.State.Running == nil
+				break
 			}
-
-			return
-		}, annotations...)
-		if err != nil {
-			return xerrors.Errorf("cannot update pod lifecycle independent state: %w", err)
+		}
+		if terminated {
+			go m.finalizeWorkspaceContent(ctx, wso)
```

Comment:
Do I understand correctly that we now wait for the root process to terminate before we start the backup process? What are the implications on stopping performance?

Reply:
It's been like this before as well - which is what made the PLIS necessary in the first place. Prior to finalizers we didn't have a way to keep the pod arround once the container was stopped, hence introduced the PLIS.